README.md
acid-base-solutions
===================

'Acid-Base Solutions' simulation

By PhET Interactive Simulations
http://phet.colorado.edu/

NOTE: This repository has dependencies on other repositories. Those repositories
much be checked out as siblings to this repository in your directory structure.
See package.json for a list of dependencies.

Third-Party Dependencies
=============

This repository uses third-party libraries.
Those libraries and their licenses are available in: https://github.com/phetsims/sherpa.



changes.txt
10/1/14 use new JSdoc conventions for @param
9/8/14 use {Object} instead of {*} as type expression for options param
8/1/14 restore CM in credits
6/26/14 #125 replace explicit prototype assignment with inherit(Object,...)
# 1.1 branched 6/19/14
# 1.1.0-dev.1 6/17/14
6/17/14 #122 rename 'Custom Solution' screen to 'My Solution'
5/27/14 change 'imports' to 'modules' in require comments
4/17/14 #119 create icon for Introduction screen programmatically, so that it looks good at all sizes
4/16/14 #111 don't show 'hand' cursor over disabled 'Solvent' check box (sun#65)
4/16/14 #109 show 'N.NN x 10^0' as 'N.NN'
4/16/14 #108 add RSC acknowledgement to About dialog
4/16/14 #113 fix 'Strong Base (MOH)' label on radio button in 'Solutions' control panel
4/16/14 #114 increase size of '+' and '-' on MOH molecule icon
# 1.0 branched 4/15/14
4/15/14 remove nitroglycerin from package.json, not used
# 1.0.0-dev.9 4/14/14
4/14/14 #105 fixes to magnifier view
4/14/14 #105 first pass at drawing deltas in magnifier view
4/11/14 joist#120 added dependency on 'brand' repository
4/10/14 #106 reduce size of solvent.png file instead of scaling it down, saves 600K, sacrifices a bit of quality but the image is shown with reduced opacity
4/10/14 #106 increase size of solvent.png and scale it down, so that H2O molecules are the proper size
4/10/14 #104 workaround for Firefox graphics acceleration bug that was manifesting itself in reaction equation arrows
4/9/14 delete unused strings, remove trailing space in 'Introduction '
4/9/14 remove dependency on nitroglycerin
4/9/14 rename ConductivityTester.location to ConductivityTester.bulbLocation
# 1.0.0-dev.8 4/9/14
4/9/14 remove dependency on proxy concentration and strength properties in graph, wire up directly to selected solution
4/9/14 delete Beaker.containsPoint convenience function
4/9/14 #101 make conductivity probes move together
4/9/14 #102 restructure the 'solution' type hierarchy
4/9/14 #99 require subtypes to provide initial values for all properties of AqueousSolution supertype
4/9/14 #99 add missing initializer for 'solute concentration' of 'Weak Acid'
4/9/14 #100 add a mask so that pH paper doesn't appear in magnifier
4/8/14 #97 fix molecule icons that were broken by previous change
4/8/14 #97 simplify implementation of 'Water (H2O)' label on radio button
4/8/14 #97 factor out ugly composition of button labels in 'Solutions' control panel
# 1.0.0-dev.7 4/7/14
4/7/14 #97 remove all scenery.HTMLText from 'Solutions' control panel, it was causing out-of-memory problems
4/7/14 #81 new artwork for light bulb, new implementation of LightBulbNode required
4/7/14 #98 Conductivity tester's brightness property should be zero when the circuit is open, delete redundant isClosed property
4/7/14 create subtypes for each screen
4/7/14 #96 create the 'Custom Solution' screen icon programmatically, so that molecule colors are correct
4/7/14 #5 improve quality of molecule images in magnifier; store at 2x their needed size, scale down when drawing to canvas
# 1.0.0-dev.6 4/6/14
4/6/14 copy model.txt and implementation-notes.txt from Java version, modify as needed
4/6/14 #5 position molecules at the centers of their images, reduce drawing region by half the lineWidth used to stroke the lens
4/6/14 change computation of magnifier radius
4/6/14 #5 scale and crop solvent.png, reduce its size from 730K to 128K
4/6/14 #5 add debug flag to show origin of magnifier
4/6/14 #5 delete common/view/magnifier/*, replaced with direct-to-Canvas implementation
4/6/14 #5 add debug flag to turn off clipping, to see how molecules are being positioned inside lens
4/6/14 #5 generate the set of molecule images needed by iterating over all solutions and their molecules
4/6/14 #5 first pass at drawing large numbers of molecules directly to canvas (no clipping)
4/6/14 #5 wire up MagnifierNode to the solution that is currently selected
4/6/14 #5 add clipping for solvent and molecules to MagnifierNode
4/6/14 draw beaker icon programmatically, delete image file
4/6/14 #94 restore concentration observer that I accidentally removed from CustomSolutionModel
4/6/15 #95 draw reaction equation arrows programmatically, so they don't look fuzzy/jagged/shifted
4/4/14 #94 back out change that keeps concentration of all solutions synchronized, since this may have performance implications
4/4/14 #94 keep strength of weak solutions synchronized at all times
4/4/14 #94 make strength slider operate on weak solutions only; never change strength of strong solutions
4/4/14 #94 delete unused isValidStrength property throughout; validate all attempts to change strength
4/4/14 #94 fix implementation of isValidStrength for strong acid and strong base
4/4/14 adjust pH meter location so that it doesn't go behind graph
4/4/14 clean up ReactionEquationFactory
# 1.0.0-dev.5 4/3/14
4/3/14 #92 reset pH paper when strength or concentration are changed in 'Custom Solution' screen
4/3/14 #86 compute color of pH paper by interpolating between colors used in pH color key
4/3/14 rename ABSControlPanel to AcidBaseSolutionsControlPanel
4/3/14 #51 factor out common stuff for main control panel into ABSControlPanel
4/3/14 #51 move 'Reset All' button to right of 'Tools' panel
4/3/14 #74 expand touch areas for all radio buttons and check boxes, increase vertical spacing in 'Solutions' and 'Views' panels
4/3/14 fix graph on 'Custom' screen so that it updates, remove vestigial viewMode and toolMode args from ConcentrationGraph constructor calls
4/3/14 #88 tweak conductivity probe locations to account for smaller beaker size
4/3/14 #88 make beaker 10% smaller, reposition it
4/3/14 #88 make reaction equations 50% larger
4/3/14 #83 change OH color to rgb(90,90,255), so that text is more visible on graph bars
4/3/14 #82 reset view properties
4/3/14 #74 expand touch area of conductivity probes
4/3/14 #91 constrain probes so that they can only be dragged part-way into the solution
4/3/14 #90 increase size of conductivity probes

js/common/ABSColors.js
// Copyright 2002-2014, University of Colorado Boulder

/**
 * Colors for simulation 'Acid-Base Solutions' simulation.
 *
 * @author Chris Malley (PixelZoom, Inc.)
 */
define( function( require ) {
  'use strict';

  // modules
  var Color = require( 'SCENERY/util/Color' );

  // constants
  var GRAY_MOLECULE = 'rgb(120,120,120)';

  return {

    SCREEN_BACKGROUND: 'white',

    CONTROL_PANEL_BACKGROUND: 'rgb(208,212,255)',

    PH_PAPER: 'rgb(217,215,154)', // color of blank pH paper, cream

    // pH colors, ordered from pH value 0-14, type Color because we'll be interpolating
    PH: [
      new Color( 182, 70, 72 ),
      new Color( 196, 80, 86 ),
      new Color( 213, 83, 71 ),
      new Color( 237, 123, 83 ),
      new Color( 246, 152, 86 ),
      new Color( 244, 158, 79 ),
      new Color( 243, 160, 78 ),
      new Color( 244, 182, 67 ),
      new Color( 231, 201, 75 ),
      new Color( 93, 118, 88 ),
      new Color( 30, 92, 89 ),
      new Color( 34, 90, 105 ),
      new Color( 39, 87, 111 ),
      new Color( 27, 67, 90 ),
      new Color( 0, 34, 52 )
    ],

    // The field names here must correspond to the 'key' fields in AqueousSolution.molecules.
    MOLECULES: {
      A: 'rgb(0,170,255)',
      B: GRAY_MOLECULE,
      BH: 'rgb(255,170,0)',
      H2O: 'rgb(164,189,193)',
      H3O: 'rgb(255,85,0)',
      HA: GRAY_MOLECULE,
      M: 'rgb(255,170,0)',
      MOH: GRAY_MOLECULE,
      OH: 'rgb(90,90,255)'
    }
  };
} );
